src/main/java/net/neoforged/neoforge/attachment/AttachmentHolder.java

```diff
@@ -99,6 +113,15 @@ protected final void deserializeAttachments(CompoundTag tag) {
      * @return {@code true} if the attachments are compatible, {@code false} otherwise
      */
     public static <H extends AttachmentHolder> boolean areAttachmentsCompatible(H first, H second) {
+        // Confirm both maps are present or not present
+        if(first.attachments == second.attachments) {
+            return true;
+        }
+        // If either of the maps are not present, then only one attachment holder has a populated map.
+        // They are not compatible.
+        if(first.attachments == null || second.attachments == null) {
```

Comment:
That's not correct. One map may have the default instance of an attachment and should therefore be compatible with a `null` map. Maybe store each map or else `Map.of()` if null in local fields.